File: apps/ios/Sources/Core/Push/PushRouteGate.swift
import Foundation

// Meloming PushNotificationManager single dispatch, adapted to the existing strict
// route queue. The provider payload is a hint, never proof of room membership.
struct PushRouteGate: Sendable {
    let parser: ContentRouteParser
    private var queue = PendingRouteQueue()
    private var scope: PushScope?
    init(parser: ContentRouteParser) { self.parser = parser }
    mutating func bind(_ value: PushScope?) {
        if scope != value { queue.resetScope(); scope = value }
    }
    mutating func offer(url: String, eventID: String, expected: PushScope, now: UInt64) -> Bool {
        guard expected == scope, let hint = parser.parse(url) else { return false }
        return queue.offer(hint, eventID: eventID, now: now, expectedScope: queue.scopeToken)
    }
    mutating func begin(now: UInt64) -> RouteTicket? { scope == nil ? nil : queue.begin(now: now) }
    // Call after real /me + room detail authorization, then navigate in this actor turn.
    mutating func authorized(_ ticket: RouteTicket, expected: PushScope, roomID: String, now: UInt64) -> RoomRouteHint? {
        guard expected == scope, ticket.hint.roomID == roomID else { return nil }
        return queue.consume(ticket, now: now)
    }
    mutating func denied(_ ticket: RouteTicket) { queue.cancel(ticket) }
}

File: apps/android/app/src/main/java/chat/rogi/rogichat/core/push/PushRouteGate.kt
package chat.rogi.rogichat.core.push

import chat.rogi.rogichat.core.navigation.ContentRouteParser
import chat.rogi.rogichat.core.navigation.PendingRouteQueue
import chat.rogi.rogichat.core.navigation.RoomRouteHint
import chat.rogi.rogichat.core.navigation.RouteTicket

/** Warm taps and cold launch share one queue, adapted from Meloming single dispatch.
 * Never trust provider payload membership; OS owner must re-fetch /me and room detail.
 */
class PushRouteGate(private val parser: ContentRouteParser) {
    private val queue = PendingRouteQueue()
    private var scope: PushScope? = null
    fun bind(value: PushScope?) {
        if (scope != value) { queue.resetScope(); scope = value }
    }
    fun offer(url: String, eventId: String, expected: PushScope, now: Long): Boolean {
        if (expected != scope) return false
        val hint = parser.parse(url) ?: return false
        return queue.offer(hint, eventId, now, queue.scopeToken)
    }
    fun begin(now: Long): RouteTicket? = if (scope != null) queue.begin(now) else null
    /** Call synchronously after real authorization returns, before navigating on same dispatcher. */
    fun authorized(ticket: RouteTicket, expected: PushScope, roomId: String, now: Long): RoomRouteHint? {
        if (expected != scope || ticket.hint.roomId != roomId) return null
        return queue.consume(ticket, now)
    }
    fun denied(ticket: RouteTicket) { queue.cancel(ticket) }
}
